src/services/refactors/extractType.ts

```diff
@@ -0,0 +1,141 @@
+    interface TypedefInfo { isJS: true; selection: TypeNode; firstStatement: Statement & HasJSDoc; }
+    type Info = TypeAliasInfo | TypedefInfo;
+    function getRangeToExtract(context: RefactorContext): Info | undefined {
+        const { file, startPosition } = context;
+        const isJS = isSourceFileJS(file);
+        const current = getTokenAtPosition(file, startPosition);
+        const range = createTextRangeFromSpan(getRefactorContextSpan(context));
+
+        const selection = findAncestor(current, (node => node.parent && rangeContainsSkipTrivia(range, node, file) && !rangeContainsSkipTrivia(range, node.parent, file)));
+        if (!selection || !isTypeNode(selection)) return undefined;
+
+        if (isJS) {
+            // typeparam tag is not supported yet
+            return {
+                isJS,
+                selection,
+                firstStatement: Debug.assertDefined((findAncestor(selection, isStatementButNotBlockAndHasJSDoc))),
+            };
+        }
+        else {
+            const firstStatement = Debug.assertDefined((findAncestor(selection, isStatementButNotBlock)));
+            const typeParameters = checkAndCollectionTypeArguments(context.program.getTypeChecker(), selection, firstStatement, file);
+            if (!typeParameters) return undefined;
+
+            return { isJS, selection, firstStatement, typeParameters };
+        }
+    }
+
+    function isStatementButNotBlock(n: Node): n is Statement {
+        return n && isStatement(n) && !isBlock(n);
+    }
+
+    function isStatementButNotBlockAndHasJSDoc(n: Node): n is (Statement & HasJSDoc) {
+        return isStatementButNotBlock && hasJSDocNodes(n);
+    }
+
+    function rangeContainsSkipTrivia(r1: TextRange, node: Node, file: SourceFile): boolean {
+        return rangeContainsStartEnd(r1, skipTrivia(file.text, node.pos), node.end);
+    }
+
+    function checkAndCollectionTypeArguments(checker: TypeChecker, selection: TypeNode, statement: Statement, file: SourceFile): string[] | undefined {
+        let hasError = false;
+        const result: string[] = [];
+        visitor(selection);
+        return hasError ? undefined : result;
+
+        function visitor(node: Node) {
+            if (isTypeReferenceNode(node)) {
+                if (isIdentifier(node.typeName)) {
+                    const symbol = checker.resolveName(node.typeName.text, node.typeName, SymbolFlags.TypeParameter, /* excludeGlobals */ true);
+                    if (symbol) {
+                        const declaration = first(symbol.declarations);
+                        if (rangeContainsSkipTrivia(statement, declaration, file) && !rangeContainsSkipTrivia(selection, declaration, file)) {
+                            result.push(node.typeName.text);
+                        }
+                    }
+                }
+            }
+            else if (isInferTypeNode(node)) {
+                const conditionalTypeNode = findAncestor(node, isConditionalTypeNode);

```

Comment:
i think this might be wrong. we are assuming that the infer type node is only allowed in the extends clause of a conditional type. so we want to check if `selection` includes the conditional type node to whose extends clause the infer type node belongs. this conditional type node might not be the first conditional type ancestor we find because there could be nested conditional type nodes. for example:
```ts
type Crazy<T> = T extends [infer P, (infer R extends string ? string : never)] ? P & R : string;
```
we can apply the refactoring and extract `(infer R extends string ? string : never)` to obtain this:
```ts
type NewType = (infer R extends string ? string : never); // error: 'infer' declarations are only permitted in the 'extends' clause of a conditional type

type Crazy<T> = T extends [infer P, NewType] ? P & R : string;
```
which gives an error.

Reply:
sorry, mis operation😅

Reply:
Find type node which is the first condition that contains selected infer type node in extends node